### PR TITLE
gromacs: add nvshmem variant

### PR DESCRIPTION
**Pull request by snehring**
I was experimenting with the nvshmem feature and wanted it in the spack recipe.

The cufftmp variant with presumably anything other than nvhpc doesn't detect the nvshmem libraries out of the box. This seems like an issue with the cmake routines for this feature in gromacs itself than spack, imo but modifying the environment to include paths to these libraries solves it as is. An issue that probably should be addressed in another PR is the nvhpc package, it doesn't make all the libraries (the nvshmem ones notably) available at build time or in the environment when loaded.

Also made cuda_arch actually use provided arches instead of doing default behavior if provided.

### CODE REVIEW

**Review by tldahlgren — APPROVED** [review]
LGTM. But will defer to at least one of the maintainers: @eirrgang @junghans @mabraham

**Review by junghans — APPROVED** [review]

**Comment by mabraham** [comment]
I proposed a missing conflict at https://github.com/spack/spack/pull/46065

**Pull request merged**
